client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java

```diff
@@ -75,24 +83,48 @@ public void onSuccess(int chunkIndex, ManagedBuffer buffer) {
 
           @Override
           public void onFailure(int chunkIndex, Throwable e) {
-            String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
-            logger.error(errorMsg, e);
-            exception.set(new IOException(errorMsg, e));
+            PartitionLocation failedLocation = ((ChunkClientException) e).getLocation();
+            readableLocations.remove(failedLocation);
+            if (readableLocations.isEmpty()) {
+              String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
+              logger.error(errorMsg, e);
+              exception.set(new IOException(errorMsg, e));
+            } else {
+              try {
+                synchronized (this) {
+                  client.set(
+                      new ChunkClient(
+                          conf,
+                          shuffleKey,
+                          location.getPeer(),
+                          this,
+                          clientFactory,
+                          startMapIndex,
+                          endMapIndex));
+                  currentChunkIndex.set(0);
```

Comment:
There can be concurrent issues when in-flight fetches return.

Reply:
Those variables "currentChunkIndex, returnedChunks, numChunks" are synced and atomic.